Add Linux QSV Implementation

kkartaltepe opened the pull request:
### Description
<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->
This is built on top of #8014, so the first 3 commits are from there and enables the non-texture encoding path for linux.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open GitHub Issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->
ffmpeg's vaapi encoding implementation seems to not be making much progress in av1, so we can enable it through this. This also helps reduce confusion as the typical QSV UI on windows is now applicable on linux as well.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->
Tested on a750 and a i3-7100U, there are some parts of mfx/vpl that seem broken on va-api and have been marked windows only but the largely everything works as it does on windows.

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
- New feature (non-breaking change which adds functionality)
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

kkartaltepe commented:
For now building on CI is disabled for this in linux, since I couldnt immediately figure out how to only build on 22.04 but I think it should build there. On 20.04 libmfx is too old to support AV1 definitions and we dont have any defines to disable these on older SDKs.

ma3uk commented:
What do you think about migrating from MSDN to OneVPL with the replacement of API and dispatcher files, as well as using OneVPL sessions (mfxLoad and mfxSession) for processors 11 gen and higher?
It is also possible to use internal memory management for generation 11 and above.
The main reason for switching to oneVPL is full support for the AV1 codec with support for Grain Film and other innovations that will no longer appear in MSDK.
In addition to switching to oneVPL, I think it would be great to make a field for entering additional user parameters like x264 or ffmpeg, because there are many parameters that have a positive effect on the quality of the output image, for example, you can already tell the codec that the video is NOISY, which increases the power of deblock, and in the release of OneVPL 2.9 will appear support for tuning profiles like x264 (SSIM, PSNR, VMAF and others)

ma3uk commented:
And do you need to update the code for outdated hardware? Why is it not possible to make a new implementation that will be available only on new hardware and name it as Version 3, as was done with the NVENC implementation.

tytan652 commented on CI/linux/01_install_dependencies.sh:
```diff
@@ -90,7 +90,8 @@ install_dependencies() {
          libluajit-5.1-dev python3-dev libx11-dev libxcb-randr0-dev libxcb-shm0-dev libxcb-xinerama0-dev \
          libxcb-composite0-dev libxinerama-dev libxcb1-dev libx11-xcb-dev libxcb-xfixes0-dev swig libcmocka-dev \
-         libpci-dev libxss-dev libglvnd-dev libgles2-mesa libgles2-mesa-dev libwayland-dev libxkbcommon-dev"
+         libpci-dev libxss-dev libglvnd-dev libgles2-mesa libgles2-mesa-dev libmfx-dev libwayland-dev \
+         libxkbcommon-dev"
```
This change should in the `plugin-deps` line.

tytan652 commented on cmake/Modules/FindMfx.cmake:
```diff
@@ -0,0 +1,53 @@
+# * Try to find libvpl/libmfx, prefers vpl
```
If it's two different library. Whatever their similarity, it should two different finders.

kkartaltepe replied:
These are two different dispatchers for the same API, as such either provides the functionality we need. This was an mainly an attempt to make it work on ubuntu 20, which doesnt provide the vpl dispatcher. However since ubuntu 20 doesnt provide av1 definitions it doesnt end up working. So we could remove the mfx fallback as ubuntu 22 provides vpl.

kkartaltepe commented:
> What do you think about migrating from MSDK to OneVPL with the replacement of API and dispatcher files, as well as using OneVPL sessions (mfxLoad and mfxSession) for processors 11 gen and higher?
It is also possible to use internal memory management for generation 11 and above (this will get rid of dependencies for Windows and Linux platforms (common_directx 9 and 11)).
The main reason for switching to oneVPL is full support for the AV1 codec with support for Grain Film and other innovations that will no longer appear in MSDK.

the majority of this code is Intel's and they largely support the windows implementation. I have no desire to support windows so the changes to shared code as kept as minimal as possible and avoid undue burden on the windows side.

> And do you need to update the code for outdated hardware? Why is it not possible to make a new implementation that will be available only on new hardware and name it as Version 3, as was done with the NVENC implementation.

Im not sure I understand your concern, are you having issues with this code on newer hardware? It works on gen12 for me and should work on anything back to broadwell, but i was only able to test back to KBL since thats what I have.

RytoEX commented:
> What do you think about migrating from MSDK to OneVPL with the replacement of API and dispatcher files, as well as using OneVPL sessions (mfxLoad and mfxSession) for processors 11 gen and higher?

We are planning to implement OneVPL at some point. That is not the scope of this PR. The scope and nature of the oneVPL implementation has not yet been determined. Regardless, let us focus on this PR's scope.

tytan652 commented:
Just a question that come up in my mind, does QSV on Linux have the same "non constant" rate control availability (only CQP because firmware) as VAAPI ?

If this is the case, the QSV simple preset should not be available on Linux.

kkartaltepe commented:
> Just a question that come up in my mind, does QSV on Linux have the same "non constant" rate control availability as VAAPI (only CQP because Intel firmware kernel parameter) ?
> 
> If this is the case, the QSV simple preset should not be available on Linux.

It is limited by the vaapi driver, yes. So you have the same limitations with the free/non-free driver.

norihiro commented:
Would you revise the commit message of 1c1bf020c like below? I fixed line breaks, changed linux to Linux, and fixed a typo.

```plain
obs-qsv11: Add adapter checks for Linux

This moves the existing adapter checking into the platform layer and
moves the Windows implementation to its platform implementation and adds
a Linux implementation based on directly querying VA-API.

Unlike Windows, this check takes ~1ms so we have no need to spin out
another thread to perform the work. This also fixes up some of the cpp/c
mixing going on in common_utils."h" to allow us to call common functions
from c files.
```

tytan652 reviewed (changes requested):
> It is limited by the vaapi driver, yes. So you have the same limitations with the free/non-free driver.

In case it was not understood, `SIMPLE_ENCODER_QSV` (H264 or AV1) should not be exposed to the user under Linux since there is no guarantee that the iGPU (or GPU) supports CBR which is a requirement for the simple preset in streaming.

tytan652 commented on plugins/obs-qsv11/common_utils_linux.cpp:
```diff
@@ -0,0 +1,274 @@
+		break;
+	}
+
+	return (rc & VA_RC_CBR || rc & VA_RC_CQP || rc & VA_RC_VBR);
```
Another idea:
Make sure that the adapter handle the three rate control, which allow to keep QSV in simple mode on Linux since "problematic" CPU are removed from the equation.
```suggestion
	return (rc & VA_RC_CBR && rc & VA_RC_CQP && rc & VA_RC_VBR);
```

kkartaltepe replied:
This is being used for support of the encoder in general, setting it this way would just disable support entirely on free drivers. I think it probably makes sense to not have the encoder in streaming if it doesnt support CBR, but I would prefer to keep it available for recording even in that case.

kkartaltepe replied:
(and also the h264 version of this function isnt being used at all, since the shared windows code just assumes h264 support)

BrentonPoke commented:
How far along is this one? Would it allow the use of the ARC line of GPUs?

kkartaltepe commented:
> How far along is this one? Would it allow the use of the ARC line of GPUs?

Its done, it requires some changes to windows which I don't have to test so it has been a bit slow to merge. I will be rebasing these PRs soon(tm). But it works great with my arc card and an older laptop cpu, you can build it yourself and test and if something isnt working id love to know.

kkartaltepe commented:
> > It is limited by the vaapi driver, yes. So you have the same limitations with the free/non-free driver.
> 
> In case it was not understood, `SIMPLE_ENCODER_QSV` (H264 or AV1) should not be exposed to the user under Linux since there is no guarantee that the iGPU (or GPU) supports CBR which is a requirement for the simple preset in streaming.

Right now there is no way to communicate this, we could simply disable this in the UI code linux for now. But the same encoder is used for simple/non-simple.

tytan652 commented:
> we could simply disable this in the UI code linux for now

This is what I mean, if it can't guarantee CBR all the time it must be advanced mode only.

BrentonPoke commented:
> > How far along is this one? Would it allow the use of the ARC line of GPUs?
> 
> Its done, it requires some changes to windows which I don't have to test so it has been a bit slow to merge. I will be rebasing these PRs soon(tm). But it works great with my arc card and an older laptop cpu, you can build it yourself and test and if something isnt working id love to know.

I tried cloning your fork, but am having issues regarding conflicting dependencies and OpenGL not being found (??). I'm building on Fedora and would love to be able to verify it for you but am unsure where to go from here.

kkartaltepe commented:
> I tried cloning your fork, but am having issues regarding conflicting dependencies and OpenGL not being found (??). I'm building on Fedora and would love to be able to verify it for you but am unsure where to go from here.

https://github.com/obsproject/obs-studio/wiki/Build-Instructions-For-Linux the wiki for building is the best place to figure out how to build the project.

BrentonPoke commented:
> > I tried cloning your fork, but am having issues regarding conflicting dependencies and OpenGL not being found (??). I'm building on Fedora and would love to be able to verify it for you but am unsure where to go from here.
> 
> https://github.com/obsproject/obs-studio/wiki/Build-Instructions-For-Linux the wiki for building is the best place to figure out how to build the project.

That's what I was following, I just don't know what to do about this:

```
CMake Warning at CMakeLists.txt:15 (project):
  VERSION keyword not followed by a value or was followed by a value that
  expanded to nothing.


-- The C compiler identification is GNU 12.3.1
-- The CXX compiler identification is GNU 12.3.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find OpenGL (missing: OPENGL_opengl_LIBRARY OPENGL_glx_LIBRARY
  OPENGL_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake/Modules/FindOpenGL.cmake:443 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  deps/glad/CMakeLists.txt:3 (find_package)
```

BrentonPoke commented:
I was able to resolve the dependency problem, but now it's asking me for a version for `WRITE_BASIC_CONFIG_VERSION_FILE()`
![image](https://github.com/obsproject/obs-studio/assets/6861851/dc532ab8-e4dd-4cc7-86ca-810553b9153b)

kkartaltepe commented:
> I was able to resolve the dependency problem, but now it's asking me for a version for WRITE_BASIC_CONFIG_VERSION_FILE()

If you need help building obs you should use our forum or discord, as its just going to be noisy for other reviewers here.

BrentonPoke commented:
> > I was able to resolve the dependency problem, but now it's asking me for a version for WRITE_BASIC_CONFIG_VERSION_FILE()
> 
> If you need help building obs you should use our forum or discord, as its just going to be noisy for other reviewers here.

At any rate, I was using the build instructions from the beginning and they need work for Fedora. Also, the formatting for the CLI commands is strange, so somebody should look at that, too.

RytoEX commented:
#8014 has been merged, and Ubuntu 20.04 has been dropped, so this can now be rebased.

kkartaltepe commented:
Rebased over the cmake changes, includes a cmake and legacy cmake that should work. And added the suggestion to hide QSV from streaming on linux, since simple recording always uses ICQ i've left it in.

RytoEX reviewed (changes requested):
Commit message nits:
* linux -> Linux
* cmake -> CMake
* qsv -> QSV
* cpp -> CPP
* c -> C
* intel -> Intel

There is also "VAAPI" written in one commit message but "VA-API" in the other. Choose one or the other.

Will try to do some test builds later.

kkartaltepe commented on CI/linux/02_build_obs.sh:
```diff
@@ -53,6 +53,10 @@ _configure_obs() {
         PIPEWIRE_OPTION="-DENABLE_PIPEWIRE=OFF"
     fi
 
+    if [ "${DISABLE_QSV}" ]; then
+        QSV_OPTION="-DENABLE_QSV11=OFF"
+    fi
```
We dont package all of the vpl runtimes and drivers needed to use QSV for flatpak, neither does flatpak. So we have to disable it there for now. We can eventually add it I think, but its like 3-5 intel projects to get things working last i looked.

RytoEX replied:
I don't believe Flatpak uses these build scripts.

kkartaltepe replied:
I definitely left this in the flatpak changes, and it still needs to be there.

kkartaltepe commented on UI/window-basic-main-profiles.cpp:
```diff
@@ -839,10 +839,12 @@ void OBSBasic::CheckForSimpleModeX264Fallback()
 	while (obs_enum_encoder_types(idx++, &id)) {
 		if (strcmp(id, "amd_amf_h264") == 0)
 			amd_supported = true;
+#ifdef _WIN32
 		else if (strcmp(id, "obs_qsv11") == 0)
 			qsv_supported = true;
 		else if (strcmp(id, "obs_qsv11_av1") == 0)
 			qsv_av1_supported = true;
+#endif
```
we need to ensure its available for simple mode recording and advanced mode streaming and recording. So we definitely want to register the encoder.

This is for safe fallback to x264 I think its better safe than sorry to ensure fallbacks dont accidentally choose the qsv encoders.

RytoEX replied:
I kind of don't like it because I don't want this section to become an ifdef'd mess, but I'm not sure I can suggest an alternative.

kkartaltepe replied:
fwiw intel is the only with this fractured encoder support so we shouldnt see this grow. The alternative is the fairly substantial work to expose ratecontrols up to the UI or as tytan suggested to just not load the encoder on the free drivers which basically means most users wont see this even if they could use simple recording.

RytoEX replied:
As mentioned, I don't really have an alternative to suggest, so I don't want to block on just "I don't like it".

kkartaltepe commented on UI/window-basic-settings-stream.cpp:
```diff
@@ -1652,6 +1654,7 @@ void OBSBasicSettings::ResetEncoders(bool streamOnly)
 
 	ui->simpleOutStrEncoder->addItem(ENCODER_STR("Software"),
 					 QString(SIMPLE_ENCODER_X264));
+#ifdef _WIN32
```
We must register the encoder as mentioned, this ensures its shown everywhere EXCEPT streaming, where intel's qsv may not have CBR support on linux.

kkartaltepe replied:
Resolving just so we can follow up on the other comment thread.

kkartaltepe commented on cmake/Modules/FindVpl.cmake:
```diff
@@ -0,0 +1,47 @@
+# * Try to find libvpl
```
Actually i was looking at the intel PR and theirs is named `vpl`, VPL is good with me.

kkartaltepe commented on plugins/obs-qsv11/CMakeLists.txt:
```diff
@@ -9,104 +9,115 @@ if(NOT ENABLE_QSV11)
-configure_file(cmake/windows/obs-module.rc.in obs-qsv11.rc)
-target_sources(obs-qsv11 PRIVATE obs-qsv11.rc)
-
 target_compile_definitions(obs-qsv11 PRIVATE DX11_D3D)
```
Bad merge resolution thanks

RytoEX commented:
Unfortunately, this crashes on OBS startup on my machine on Windows 11 on module load.

RytoEX commented on plugins/obs-qsv11/common_utils_windows.cpp:
```diff
@@ -125,6 +132,88 @@ void util_cpuid(int cpuinfo[4], int flags)
+void check_adapters(struct adapter_info *adapters, size_t *adapter_count)
+{
+	char *test_exe = os_get_executable_path_ptr("obs-qsv-test.exe");
+	struct dstr cmd;
+	struct dstr caps_str;
```
With this change, I get a crash on OBS startup on Windows 11. Adding back the zero-initialization resolves the crash.

kkartaltepe replied:
Thanks, i think this was just a bad rebase, but i cannot remember.

kkartaltepe commented on plugins/obs-qsv11/common_utils_linux.cpp:
```diff
@@ -0,0 +1,274 @@
+	clock_gettime(CLOCK_MONOTONIC, timestamp);
+}
+
+//void mfxInitTime();  might need this for Windows
```
These were just copied from windows unfortunately. Removed the commented code at least.

kkartaltepe commented on UI/window-basic-settings-stream.cpp:
```diff
@@ -1444,8 +1444,10 @@ static QString get_simple_fallback(const QString &enc)
 		return SIMPLE_ENCODER_AMD;
 	if (enc == SIMPLE_ENCODER_APPLE_HEVC)
 		return SIMPLE_ENCODER_APPLE_H264;
+#ifdef _WIN32
 	if (enc == SIMPLE_ENCODER_QSV_AV1)
 		return SIMPLE_ENCODER_QSV;
+#endif
```
I think this is probably ok to remove, since we already ifdef out before the fallback is invoked. Removed.

tytan652 commented on plugins/obs-qsv11/common_utils_linux.cpp:
```diff
@@ -0,0 +1,279 @@
+
+	const char *driver = vaQueryVendorString(display);
+	if (strstr(driver, "Intel i965 driver") != nullptr) {
+		blog(LOG_ERROR,
+		     "Legacy intel-vaapi-driver detected, incompatible with QSV");
```
Is this really needs to be a error ?
A warning might be better because it only shows that the adapter use the old VAAPI driver, which mean that the adapter is too old for the plugin.

kkartaltepe replied:
It does not mean the card is too old, a fairly wide selection of chips sit in the overlap of support between both drivers where users could have this functionality if they used the newer driver.

kkartaltepe commented:
Resolved comments.

> Unfortunately, this crashes on OBS startup on my machine on Windows 11 on module load.

With the dstr change fixed it was running fine on windows again?

RytoEX commented:
> Resolved comments.
> 
> > Unfortunately, this crashes on OBS startup on my machine on Windows 11 on module load.
> 
> With the dstr change fixed it was running fine on windows again?

That is my recollection from when I saw the crash and tested last. The crash stack pointed to the subsequent dstr_copy call, IIRC. Will retest as soon as I can.

RytoEX reviewed (approved):
Overall, this seems fine. Everything seems to work fine on my Windows system with an Intel Arc A770M. I'd appreciate review/testing from others, particularly on Linux.

RytoEX commented on plugins/obs-qsv11/common_utils_linux.cpp:
```diff
@@ -0,0 +1,274 @@
+#include <string.h>
+#include <dirent.h>
+
+mfxStatus simple_alloc(mfxHDL pthis, mfxFrameAllocRequest *request,
+		       mfxFrameAllocResponse *response)
```
I think that project preference is to only use `UNUSED_PARAMETER` if there is no other alternative. That said, I don't know if anyone thinks it's worth blocking on this.

RytoEX replied:
See, for example, #8167.

kkartaltepe replied:
Right sorry, my main reservation is that this makes implementing these a bit noiser. E.g. see the implementations in https://github.com/obsproject/obs-studio/commit/4867e00f94063214057e236f0b18232454a2a457

These are not unused because they will never be unused, just that this functionality is not implemented yet (in this PR). It is already in the pipe though.

RytoEX replied:
Sure, that's a valid reason to keep them as-is for now. Thanks for clarifying.

kkartaltepe commented:
Rebased and logging the disabling of QSV due to drivers as a warning.
